feat: persisted Station ID passed to Zinnia (#424)
filecoin-station/core: README.md

  - Windows: `%TEMP%/Filecoin Station Core`
- `$STATE_ROOT` _(string; optional)_: Station stores logs and module state in
  this directory. Defaults to

  - Linux: `${XDG_STATE_HOME:-~/.local/state}/filecoin-station-core`
  - macOS: `~/Library/Application Support/app.filstation.core`
  - Windows: `%LOCALAPPDATA%/Filecoin Station Core`

>   **IMPORTANT:** State files must be stored in a directory that's local to the

Reviewer: ```suggestion
  **IMPORTANT:** The`$STATE_ROOT` directory must be local to the
```

Reviewer: As a reader I wouldn't be certain whether by "State files" we mean that directory

  `0x000000000000000000000000000000000000dEaD`. Please note that any earnings
  sent there will be lost.

- `PASSPHRASE` _(string; required)_: a passphrase to protect the Station

>   instance private key stored in a file inside `STATE_ROOT` directory.

Author: Adding a new required env var is a breaking change. The next release will break all Core deployments using automatic upgrades (e.g. via Watchtower for Docker).

It's the least amount of work for us. However, I am concerned about the impact on our users.

How can we make this change backwards compatible?

One option that comes to my mind:
- If `PASSPHRASE` is not provided, then use one of the following fallbacks (we should pick & implement only one of them):
  1. Use the FIL_WALLET_ADDRESS as the passphrase
  2. Use a hard-coded passphrase
  3. Don't encrypt the private key at all
- When persisting the keypair, add a flag indicating whether we used the fallback or the user-provided passphrase.
- When loading the keypair, check the flag to decide how to decrypt the persisted keypair.
- When loading the keypair encrypted using the fallback mechanism when the user provided the passphrase, then upgrade our secret store to encrypt the keypair using the provided passphrase.
- When loading the keypair using the fallback mechanism, print a deprecation warning asking the user to configure their own PASSPHRASE.

It's a non-trivial amount of work to achieve backwards compatibility. Is it worth it? Can we find a simpler way?

Reviewer 2: Instead of flagging the use of a fallback versus user-provided passphrase, can we treat them as one and the same, and provide a way for the user to update their passphrase at any time?

Then we can remove these two bullet points of work:
~When persisting the keypair, add a flag indicating whether we used the fallback or the user-provided passphrase.~
~When loading the keypair, check the flag to decide how to decrypt the persisted keypair.~

And we print the deprecation warning (if we want to at all) based on if the passphrase (hash) matches the fallback (hash) that we chose

Author: > provide a way for the user to update their passphrase at any time

Interesting! How do you propose to implement this? A new sub-command for station core? E.g. `station edit-passphrase`? That could work 🤔

Reviewer 2: That's what I was imagining, although I haven't used core so I didn't know the specifics enough to mention it as a suggestion.

What do we want to do for Desktop then? Do we need UI changes to allow this functionality? Or should we just use keytar for Desktop deployments, since it already uses it?

Author: > What do we want to do for Desktop then? Do we need UI changes to allow this functionality? Or should we just use keytar for Desktop deployments, since it already uses it?

I discussed this with Julian, and we are thinking of deriving a passphrase from the wallet seed. We could add another keychain item via keytar, but that would add another prompt on macOS, which I don't like.

The idea is to derive the passphrase in such a way that the passphrase is unique enough & difficult to guess but also does not reveal too much of the seed phrase to the Station Core, in case there is a security vulnerability there.

Reviewer: > Adding a new required env var is a breaking change. The next release will break all Core deployments using automatic upgrades (e.g. via Watchtower for Docker).
>
> It's the least amount of work for us. However, I am concerned about the impact on our users.

What about making it optional instead? It's also optional with ssh. Optional but advised

When a passphrase isn't used, the private key is not encrypted at all. When you start passing one, it will fail and tell the user to delete the private key located at PATH, if they want to set a new passphrase.

This way we achieve backwards compatibility, give users an easy upgrade path, and just have 1 key regeneration per node max.

Author: > > Adding a new required env var is a breaking change. The next release will break all Core deployments using automatic upgrades (e.g. via Watchtower for Docker).
> > It's the least amount of work for us. However, I am concerned about the impact on our users.
> 
> What about making it optional instead? It's also optional with ssh. Optional but advised
> 
> When a passphrase isn't used, the private key is not encrypted at all. When you start passing one, it will fail and tell the user to delete the private key located at PATH, if they want to set a new passphrase.
> 
> This way we achieve backwards compatibility, give users an easy upgrade path, and just have 1 key regeneration per node max.

That's an option, too.

When the user regenerates the key, the Station ID will change. We will count one station instance as two instances in that month. That's why I prefer to implement an automatic update of the keypair file when the passphrase changes.

If we are fine with the possibility that Station ID can infrequently change and thus we will incorrectly count the number of unique instances, then I am proposing a small change to your proposal:
- When the user starts providing a passphrase, we automatically delete the old file and generate a new one.
- We can broaden the conditions triggering new keypair creation to all cases where the file cannot be decrypted using the provided passphrase. This way, the Station Core operator does not need to perform any manual steps related to the passphrase. The downside is that whenever the passphrase changes, so does the Station ID, and therefore, we can double-count station instances more often than we would like to.

Reviewer: > When the user regenerates the key, the Station ID will change. We will count one station instance as two instances in that month. That's why I prefer to implement an automatic update of the keypair file when the passphrase changes.

Yes, but we only have 1 double counting per node max, one time. I don't think that's particularly bad and worth more complexity

Reviewer: > When the user starts providing a passphrase, we automatically delete the old file and generate a new one.

How do we detect when a user starts providing a passphrase? We can only tell that a passphrase was provided, but not that one wasn't provided before, right? Or can we deduct this information from the file?

If so, what about instead of regenerating the private key, we simply take the unencrypted value, encrypt it, and overwrite the file? This way the operator doesn't notice and the station id doesn't change.

Reviewer: If this automatic process isn't possible, I'm fully in favor of making the core operator manage this. They're most likely a linux admin level user.

Author: I think it's easy to detect if the file is not encrypted - it will contain plaintext JSON starting with the character `{`. The encrypted file contains binary data.

I'll update this PR with the following changes:
- Make the PASSPHRASE optional
- When PASSPHRASE is not provided, we store the keypair in plaintext
- When loading the keypair, handle the upgrade path from unencrypted to encrypted

Author: > I think it's easy to detect if the file is not encrypted - it will contain plaintext JSON starting with the character `{`. The encrypted file contains binary data.
> 
> I'll update this PR with the following changes:
> 
> * Make the PASSPHRASE optional
> * When PASSPHRASE is not provided, we store the keypair in plaintext
> * When loading the keypair, handle the upgrade path from unencrypted to encrypted

Implemented in [a9babf9](https://github.com/filecoin-station/core/pull/424/commits/a9babf9e8a44442e9f13bec51cf780d4c80347f2)
